{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Report generation example codes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Show the metadata as a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.graph_objects as go\n",
    "import os\n",
    "\n",
    "## ---------------------------------------------------\n",
    "## Load the meta data, in csv, tsv, or excel format  -\n",
    "## ---------------------------------------------------\n",
    "meta_file = \"meta.csv\" # Replace with your file path\n",
    "data = pd.read_csv(meta_file)\n",
    "\n",
    "## ------------------------------------\n",
    "## The following section is optional  -\n",
    "## ------------------------------------\n",
    "# sort data first by Serotype and then by Accession\n",
    "data = data.sort_values(by=['Serotype', 'Date of Collection']) # Example columns; You may sort by other columns\n",
    "#replace Nan with empty string\n",
    "data = data.fillna('')\n",
    "# rename column \"Origin\" to \"Country of Origin\" as example\n",
    "data = data.rename(columns={\"Origin\": \"Country of Origin\"})\n",
    "\n",
    "\n",
    "## ----------------------------------------------------\n",
    "## Plot the data in a table, details are configurable -\n",
    "## ----------------------------------------------------\n",
    "fig = go.Figure(data=[go.Table(\n",
    "    header=dict(values= list(data.keys()), \n",
    "                align='center',\n",
    "                font=dict(color='black', size=14),\n",
    "                ),\n",
    "    # show two places after decimal\n",
    "    cells=dict( values= [\n",
    "                                data[col].round(2) if data[col].dtype in ['float64', 'int64'] else data[col]\n",
    "                                for col in data],\n",
    "                fill_color='lavender',\n",
    "                align='center',\n",
    "                font=dict(color='black', size=14),\n",
    "                height=25,\n",
    "            ),\n",
    "            # columnwidth=[80, 80, 80, 80, 80, 80], # you may adjust the column width depending on the data\n",
    ")])\n",
    "\n",
    "fig.update_layout(margin=dict(l=30, r=30, t=30, b=30))\n",
    "fig.update_layout(width=1500, height=1000) # You may change the width and height to better display the data\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Plot phylogenetic tree as a whole tree given a newick format file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "import pandas as pd\n",
    "from plotly.offline import init_notebook_mode, iplot\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "init_notebook_mode(connected=False)\n",
    "import colorsys\n",
    "import numpy as np\n",
    "import os\n",
    "import random\n",
    "\n",
    "\n",
    "\n",
    "def get_max_depth(tree):\n",
    "    terminals = tree.get_terminals()\n",
    "    depths_from_root = {terminal.name: tree.distance(tree.root, terminal) for terminal in terminals}\n",
    "    max_depth = max(depths_from_root.values())\n",
    "    return max_depth\n",
    "\n",
    "\n",
    "def get_x_coordinates(tree):\n",
    "    \"\"\"Associates to  each clade an x-coord.\n",
    "    returns dict {clade: x-coord}\n",
    "    \"\"\"\n",
    "    xcoords = tree.depths()\n",
    "    if not max(xcoords.values()):\n",
    "        xcoords = tree.depths(unit_branch_lengths=True)\n",
    "    return xcoords\n",
    "\n",
    "\n",
    "def get_y_coordinates(tree, dist=1.3):\n",
    "\n",
    "    maxheight = tree.count_terminals()\n",
    "    # Rows are defined by the tips/leafs\n",
    "    ycoords = dict((leaf, maxheight - i * dist) for i, leaf in enumerate(reversed(tree.get_terminals())))\n",
    "\n",
    "    def calc_row(clade):\n",
    "        for subclade in clade:\n",
    "            if subclade not in ycoords:\n",
    "                calc_row(subclade)\n",
    "        ycoords[clade] = (ycoords[clade.clades[0]] + ycoords[clade.clades[-1]]) / 2\n",
    "\n",
    "    if tree.root.clades:\n",
    "        calc_row(tree.root)\n",
    "    return ycoords\n",
    "\n",
    "\n",
    "def get_clade_lines(\n",
    "    orientation=\"horizontal\",\n",
    "    y_curr=0,\n",
    "    x_start=0,\n",
    "    x_curr=0,\n",
    "    y_bot=0,\n",
    "    y_top=0,\n",
    "    line_color=\"rgb(25,25,25)\",\n",
    "    line_width=0.5,\n",
    "):\n",
    "    \"\"\"define a shape of type 'line', for branch\"\"\"\n",
    "    branch_line = dict(type=\"line\", layer=\"below\", line=dict(color=line_color, width=line_width))\n",
    "    if orientation == \"horizontal\":\n",
    "        branch_line.update(x0=x_start, y0=y_curr, x1=x_curr, y1=y_curr)\n",
    "    elif orientation == \"vertical\":\n",
    "        branch_line.update(x0=x_curr, y0=y_bot, x1=x_curr, y1=y_top)\n",
    "    else:\n",
    "        raise ValueError(\"Line type can be 'horizontal' or 'vertical'\")\n",
    "\n",
    "    return branch_line\n",
    "\n",
    "\n",
    "def draw_clade(\n",
    "    clade,\n",
    "    x_start,\n",
    "    line_shapes,\n",
    "    x_coords,\n",
    "    y_coords,\n",
    "    line_color=\"rgb(25,25,25)\",\n",
    "    line_width=1,\n",
    "):\n",
    "    \"\"\"Recursively draw the tree branches, down from the given clade\"\"\"\n",
    "\n",
    "    x_curr = x_coords[clade]\n",
    "    y_curr = y_coords[clade]\n",
    "\n",
    "    # Draw a horizontal line from start to current\n",
    "    branch_line = get_clade_lines(\n",
    "        orientation=\"horizontal\",\n",
    "        y_curr=y_curr,\n",
    "        x_start=x_start,\n",
    "        x_curr=x_curr,\n",
    "        line_color=line_color,\n",
    "        line_width=line_width,\n",
    "    )\n",
    "\n",
    "    line_shapes.append(branch_line)\n",
    "\n",
    "    if clade.clades:\n",
    "        # Draw a vertical line connecting all children\n",
    "        y_top = y_coords[clade.clades[0]]\n",
    "        y_bot = y_coords[clade.clades[-1]]\n",
    "\n",
    "        line_shapes.append(\n",
    "            get_clade_lines(\n",
    "                orientation=\"vertical\",\n",
    "                x_curr=x_curr,\n",
    "                y_bot=y_bot,\n",
    "                y_top=y_top,\n",
    "                line_color=line_color,\n",
    "                line_width=line_width,\n",
    "            )\n",
    "        )\n",
    "\n",
    "        # Draw descendants\n",
    "        for child in clade:\n",
    "            draw_clade(child, x_curr, line_shapes, x_coords=x_coords, y_coords=y_coords)\n",
    "\n",
    "# These are predefined colors. Feel free to make edits to the list.\n",
    "def generate_distinct_colors(elements):\n",
    "    predefined_rgb_color = [\n",
    "        \"red\",\n",
    "        \"blue\",\n",
    "        \"green\",\n",
    "        \"darkorange\",\n",
    "        \"purple\",\n",
    "        \"brown\",\n",
    "        \"darkgoldenrod\",\n",
    "        \"deepskyblue\",\n",
    "        \"magenta\",\n",
    "        \"darkviolet\",\n",
    "        \"darkturquoise\",\n",
    "        \"darkslategray\",\n",
    "        \"darkkhaki\",\n",
    "        \"peru\",\n",
    "        \"darkred\",\n",
    "        \"darkcyan\",\n",
    "        \"goldenrod\",\n",
    "        \"lightblue\",\n",
    "        \"#bcbd22\",\n",
    "        \"#e377c2\",\n",
    "        \"#8c564b\",\n",
    "        \"#d62728\",\n",
    "        \"#2ca02c\",\n",
    "        \"#ff7f0e\",\n",
    "        \"#1f77b4\",\n",
    "    ]\n",
    "    colors = {}\n",
    "    # iterate over all elements and assign them a color\n",
    "    for i, element in enumerate(elements):\n",
    "        colors[element] = predefined_rgb_color[i]\n",
    "    return colors\n",
    "\n",
    "\n",
    "def plot_tree(\n",
    "    tree,\n",
    "    meta,\n",
    "    colors_for_counties,\n",
    "    title,\n",
    "    xaxis_title=\"Distance\",\n",
    "    intermediate_node_color=\"rgb(25,25,25)\",\n",
    "    width=1500,\n",
    "    height=2000,\n",
    "    plot_column_name=\"speciesID_mash\", # the column name from the meta data to be used for coloring, such as \"speciesID_mash\"\n",
    "):\n",
    "    max_depth = get_max_depth(tree)\n",
    "\n",
    "    x_coords = get_x_coordinates(tree)\n",
    "    y_coords = get_y_coordinates(tree)\n",
    "\n",
    "    line_shapes = []\n",
    "    draw_clade(\n",
    "        tree.root,\n",
    "        0,\n",
    "        line_shapes,\n",
    "        x_coords=x_coords,\n",
    "        y_coords=y_coords,\n",
    "        line_color=intermediate_node_color,\n",
    "        line_width=1,\n",
    "    )\n",
    "\n",
    "    my_tree_clades = x_coords.keys()\n",
    "    X = []\n",
    "    Y = []\n",
    "    text = []\n",
    "    hover_text = []\n",
    "    color = []\n",
    "    size = []\n",
    "\n",
    "    for cl in my_tree_clades:\n",
    "        X.append(x_coords[cl])\n",
    "        Y.append(y_coords[cl])\n",
    "        if cl.is_terminal():\n",
    "            name = cl.name\n",
    "            size.append(10)\n",
    "            # add color for accorsponding county according to the name\n",
    "            # the meta data file has to have a column named \"Sample ID\", or you have the change sample id column name here\n",
    "            county_name = meta.loc[meta[\"Sample ID\"] == name, plot_column_name].iloc[0]\n",
    "            color.append(colors_for_counties[county_name])\n",
    "\n",
    "        else:\n",
    "            name = \"\"\n",
    "            size.append(0)\n",
    "            color.append(intermediate_node_color)\n",
    "        text.append(name)\n",
    "        hover_text.append(name)\n",
    "\n",
    "    meta_county = meta.copy()\n",
    "    # reset the index\n",
    "    meta_county = meta_county.reset_index(drop=True)\n",
    "    for k, strain in meta_county.iterrows():\n",
    "        i = text.index(strain.iloc[0])\n",
    "        \n",
    "        ## ------------------------------------------------------\n",
    "        ## This section is to update what to show in hover text -\n",
    "        ## ------------------------------------------------------\n",
    "        hover_text[i] = (\n",
    "            hover_text[i]\n",
    "            + \"<br>SpeciesID_mash: \"\n",
    "            + str(meta_county.loc[k, \"speciesID_mash\"]) # speciesID_mash is a column in the metadata\n",
    "            + \"<br>SpeciesID_kraken: \"\n",
    "            + str(meta_county.loc[k, \"speciesID_kraken\"]) # speciesID_kraken is a column in the metadata\n",
    "            + \"<br>Date: \"\n",
    "            + str(meta_county.loc[k, \"Collection Date\"]) # Collection Date is a column in the metadata\n",
    "            + \"<br>County: \"\n",
    "            + str(meta_county.loc[k, \"County\"]) # County is a column in the metadata    \n",
    "            + \"<br>NDM genes: \"\n",
    "            + str(meta_county.loc[k, \"NDM\"]) # NDM is a column in the metadata\n",
    "        )\n",
    "        \n",
    "    axis = dict(showline=False, zeroline=False, showgrid=False, showticklabels=False, title=\"\")\n",
    "\n",
    "    nodes = dict(\n",
    "        type=\"scatter\",\n",
    "        x=X,\n",
    "        y=Y,\n",
    "        mode=\"markers+text\",\n",
    "        marker=dict(\n",
    "            color=color,\n",
    "            size=size,\n",
    "        ),\n",
    "        text=text,\n",
    "        hovertext=hover_text,\n",
    "        hoverinfo=\"text\",\n",
    "        textposition=\"middle right\",\n",
    "        textfont=dict(family=\"Arial\", size=12, color=color),\n",
    "    )\n",
    "\n",
    "    layout = dict(\n",
    "        title=title,  #'Phylogeny of Nmen Virus<br>86 genomes',\n",
    "        font=dict(family=\"Balto\", size=20),\n",
    "        width=width,\n",
    "        height=height,\n",
    "        autosize=False,\n",
    "        showlegend=False,\n",
    "        xaxis=dict(\n",
    "            showline=True,\n",
    "            zeroline=False,\n",
    "            showgrid=False,\n",
    "            ticklen=4,\n",
    "            showticklabels=True,\n",
    "            title=xaxis_title,  # \"Phylogeny of Nmen Virus\"\n",
    "        ),\n",
    "        yaxis=axis,\n",
    "        hovermode=\"closest\",\n",
    "        shapes=line_shapes,\n",
    "        plot_bgcolor=\"rgb(250,250,250)\",\n",
    "        margin=dict(l=10),\n",
    "    )\n",
    "\n",
    "    ## ----------------------------------------------------------\n",
    "    ## The following section is to update annotation positions  -\n",
    "    ## The annotation color is based on plot_column_name.       -\n",
    "    ## In this example, even though the variable is county, but -\n",
    "    ## the actual variable could be species or other variables  -\n",
    "    ## ----------------------------------------------------------\n",
    "    # Create annotations for each county at the top right corner\n",
    "    annotations = []\n",
    "    start_y = 0.90  # Starting y position, near the top of the plot\n",
    "    y_offset = 0.025 * (2500 / height)  # Vertical space between annotations\n",
    "\n",
    "    for i, county in enumerate(meta[plot_column_name].unique()):\n",
    "        annotations.append(\n",
    "            {\n",
    "                \"x\": 0.91,  # Near the left edge of the plot\n",
    "                \"y\": start_y - (i * y_offset),  # Move down for each new annotation\n",
    "                \"xref\": \"paper\",\n",
    "                \"yref\": \"paper\",\n",
    "                \"text\": county,\n",
    "                \"showarrow\": False,\n",
    "                \"font\": {\"family\": \"Arial\", \"size\": 20, \"color\": \"black\"},\n",
    "                \"xanchor\": \"left\",  # Align text to the left\n",
    "                \"align\": \"left\",\n",
    "                \"bgcolor\": colors_for_counties[county],  # Background color of annotation\n",
    "            }\n",
    "        )\n",
    "\n",
    "    # Update layout with annotations\n",
    "    layout[\"annotations\"] = annotations\n",
    "\n",
    "    fig = dict(data=[nodes], layout=layout)\n",
    "    fig[\"layout\"][\"xaxis\"].update({\"range\": [0, max_depth * 1.2]})  # x-axis range\n",
    "    return go.Figure(fig)\n",
    "\n",
    "\n",
    "def draw_annotation_line(\n",
    "    fig,\n",
    "    node_0: str,\n",
    "    node_1: str,\n",
    "    line_color: str,\n",
    "    dist_df,\n",
    "    #  max_depth:float,\n",
    "):\n",
    "    text = fig._data_objs[0].text\n",
    "    x = fig._data_objs[0].x\n",
    "    y = fig._data_objs[0].y\n",
    "    index_0 = text.index(node_0)\n",
    "    index_1 = text.index(node_1)\n",
    "\n",
    "    fig.add_shape(\n",
    "        type=\"line\",\n",
    "        x0=x[index_0],\n",
    "        y0=y[index_0],  # start position\n",
    "        x1=x[index_1],\n",
    "        y1=y[index_1],  # end position\n",
    "        line=dict(\n",
    "            color=line_color,\n",
    "            width=4,\n",
    "            dash=\"dot\",  # Make the line dashed\n",
    "        ),\n",
    "    )\n",
    "    snp_distance = dists_df[node_0][node_1]\n",
    "    if snp_distance > 1:\n",
    "        snp_distance = int(snp_distance)\n",
    "    fig.add_annotation(\n",
    "        x=(x[index_0] + x[index_1]) / 2,\n",
    "        y=(y[index_0] + y[index_1]) / 2,\n",
    "        # ax=-max_depth*0.0015,\n",
    "        # ay=0,                             # No vertical shift\n",
    "        text=f\"<b>Distance: {snp_distance}</b>\",\n",
    "        font=dict(size=18, color=line_color),\n",
    "        # showarrow=False,\n",
    "        arrowhead=2,\n",
    "        arrowsize=1,\n",
    "        arrowwidth=2,\n",
    "        arrowcolor=\"red\",\n",
    "    )\n",
    "\n",
    "\n",
    "meta_file = \"meta.csv\" # read the meta data file; replace with your meta data file\n",
    "                        # the meta data file has to have a column named \"\"Sample ID\", or you have the change \"Sample ID\" in the plot_tree function\n",
    "meta = pd.read_csv(meta_file)\n",
    "\n",
    "species = meta[\"speciesID_mash\"].unique() # in this example, the data points are colored by speciesID_mash\n",
    "\n",
    "colors_for_species = generate_distinct_colors(species)\n",
    "intermediate_node_color = \"rgb(25,25,25)\"\n",
    "\n",
    "dists_df = pd.read_csv(\"distances.tsv\", sep=\"\\t\", index_col=0) # read the pairwise distance file; replace with your pairwise distance file\n",
    "\n",
    "tree = Phylo.read(\"tree.newick\", \"newick\") # read the tree file; replace with your tree file\n",
    "\n",
    "fig = plot_tree(\n",
    "    tree,\n",
    "    meta,\n",
    "    colors_for_species,\n",
    "    title=\"Phylogeny of <i>NDM Positive CROs</i> samples<br>\" + str(len(meta)) + \" samples\", # the title of the plot\n",
    "    height=3000,\n",
    "    plot_column_name=\"speciesID_mash\",\n",
    ")\n",
    "\n",
    "## ---------------------------------------------------------------------------\n",
    "## This section is optional, it is to draw the annotation line between nodes -\n",
    "## ---------------------------------------------------------------------------\n",
    "draw_annotation_line(fig, \"ID#\", \"ID#\", \"black\", dists_df) # example line; you may change the sample ids here\n",
    "draw_annotation_line(fig, \"ID#\", \"ID#\", \"black\", dists_df) # example line; you may change the sample ids here\n",
    "draw_annotation_line(fig, \"ID#\", \"ID#\", \"black\", dists_df) # example line; you may change the sample ids here\n",
    "\n",
    "iplot(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Plot multiple sub-trees according to a column. In this example, the column is \"speciesID_mash\". Thus, one tree will be plotted for each species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "import pandas as pd\n",
    "from plotly.offline import init_notebook_mode, iplot\n",
    "\n",
    "init_notebook_mode(connected=False)\n",
    "import colorsys\n",
    "import numpy as np\n",
    "from Bio.Phylo.BaseTree import Tree\n",
    "from ete3 import Tree\n",
    "\n",
    "meta_file = \"meta.csv\" # read the meta data file; replace with your meta data file\n",
    "meta = pd.read_csv(meta_file)\n",
    "all_species = meta[\"speciesID_mash\"].unique() # in this example, the column to separate the sub-trees is \"speciesID_mash\"\n",
    "\n",
    "colors_for_species = generate_distinct_colors(species)\n",
    "intermediate_node_color = \"rgb(25,25,25)\"\n",
    "\n",
    "\n",
    "genes = meta[\"NDM\"].unique()\n",
    "genes.sort()\n",
    "colors_for_genes = generate_distinct_colors(genes) # in this example, the data points are colored by NDM in each sub-tree\n",
    "\n",
    "tree_file = \"tree.newick\" # read the tree file; replace with your tree file\n",
    "\n",
    "index = 0\n",
    "for species in all_species:\n",
    "    # obtain the list of samples in meta for the county\n",
    "    species_samples = meta[meta[\"speciesID_mash\"] == species]\n",
    "    sample_list_by_species = species_samples[\"Sample ID\"].tolist()\n",
    "    if len(species_samples) <= 1:\n",
    "        continue\n",
    "\n",
    "    tree = Tree(tree_file, format=1)  # read the whole tree\n",
    "    tree.prune(sample_list_by_species)  # prune the tree to the samples in the species\n",
    "    out_file = \"subtree.\" + species + \".nwk\"\n",
    "    tree.write(outfile=out_file, format=1)  # write the subtree to a newick file\n",
    "\n",
    "    subtree = Phylo.read(out_file, \"newick\")\n",
    "\n",
    "    title = species + \"<br>\" + str(len(species_samples)) + \" samples\" # the title of the plot\n",
    "\n",
    "    subfig = plot_tree(\n",
    "        subtree,\n",
    "        species_samples,\n",
    "        colors_for_genes,\n",
    "        title=title,\n",
    "        xaxis_title=\"Mash distance\",\n",
    "        width=1500, # the width of the plot\n",
    "        height=500 + 10*len(species_samples), # the height of the plot\n",
    "        plot_column_name=\"NDM\", # the column to be used for coloring\n",
    "    )\n",
    "\n",
    "    iplot(subfig)\n",
    "    os.remove(out_file)     # remove out_file, which is a temporary file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. 3D plot of the samples based on the pairwise distances. Either Multidimensional Scaling (MDS) or Uniform Manifold Approximation and Projection (UMAP) may be used for generating the coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.manifold import MDS\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "import umap.umap_ as umap\n",
    "from sklearn.decomposition import PCA\n",
    "import warnings\n",
    "\n",
    "# Suppress all FutureWarnings\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "warnings.filterwarnings(\"ignore\", message=\"using precomputed metric; inverse_transform will be unavailable\")\n",
    "warnings.filterwarnings(\"ignore\", message=\"n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.\")\n",
    "\n",
    "\n",
    "metadata_path = \"meta.csv\" # read the meta data file; replace with your meta data file\n",
    "metadata = pd.read_csv(metadata_path)\n",
    "species = metadata['speciesID_mash'].unique() # in this example, the column to separate the sub-trees is \"speciesID_mash\"\n",
    "colors_for_species = generate_distinct_colors(species) # in this example, the data points are colored by speciesID_mash\n",
    "\n",
    "# zip the species and colors_for_species\n",
    "species_colors = dict(zip(species, colors_for_species.values()))\n",
    "\n",
    "distance_matrix_path = \"distances.tsv\" # read the pairwise distance file; replace with your pairwise distance file\n",
    "distance_matrix = pd.read_csv(distance_matrix_path, sep=\"\\t\", index_col=0)\n",
    "distance_matrix = distance_matrix.fillna(0)\n",
    "\n",
    "distance_array = distance_matrix.values\n",
    "\n",
    "mds = MDS(n_components=3, dissimilarity='precomputed', random_state=42) # in this example, the data points are colored by speciesID_mash\n",
    "# mds = umap.UMAP(n_components=3, metric='precomputed', random_state=42) # you may also use UMAP for dimensionality reduction\n",
    "coordinates = mds.fit_transform(distance_array)\n",
    "\n",
    "fig = px.scatter_3d(\n",
    "    x=coordinates[:, 0],\n",
    "    y=coordinates[:, 1],\n",
    "    z=coordinates[:, 2],\n",
    "    color=metadata['speciesID_mash'].astype(str),\n",
    "    labels={'x': 'Dimension 1', 'y': 'Dimension 2', 'z': 'Dimension 3', 'color': 'speciesID_mash'},\n",
    "\n",
    "    # example hover data; you may change the column names here\n",
    "    hover_data={\n",
    "        'Sample ID': metadata['Sample ID'],\n",
    "        'Collection Date': metadata['Collection Date'],\n",
    "        'County': metadata['County'],\n",
    "        'speciesID_mash': metadata['speciesID_mash'].astype(str),\n",
    "        'NDM': metadata['NDM'],\n",
    "    },\n",
    "    color_discrete_map={str(k): v for k, v in species_colors.items()},\n",
    "    width=1500, # the width of the plot\n",
    "    height=1500, # the height of the plot\n",
    ")\n",
    "\n",
    "# Update hover template to exclude dimension data\n",
    "# example hover data; you may change the column names here\n",
    "fig.update_traces(\n",
    "    hovertemplate=\"<br>\".join([\n",
    "        \"Species: %{customdata[3]}\",\n",
    "        \"Sample ID: %{customdata[0]}\",\n",
    "        \"Collection Date: %{customdata[1]}\",\n",
    "        \"County: %{customdata[2]}\",\n",
    "        \"NDM: %{customdata[4]}\",\n",
    "    ]),\n",
    "    customdata=metadata[['Sample ID', 'Collection Date', 'County', 'speciesID_mash', 'NDM']].values\n",
    ")\n",
    "\n",
    "fig.update_traces(marker=dict(size=3))\n",
    "# Configure the legend\n",
    "fig.update_layout(\n",
    "    legend=dict(\n",
    "        title='Species',\n",
    "        tracegroupgap=10,\n",
    "        font=dict(size=14),\n",
    "        itemsizing='constant',\n",
    "        x=1,\n",
    "        y=0.5,\n",
    "    ),\n",
    "    font=dict(size=12)\n",
    ")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Antimicrobial Resistance Genes. The following tables outline the antimicrobial resistance (AMR) genes present in each sample. The data was generated using AMRFinderPlus (v3.10.1) included in Sanibel pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.graph_objects as go\n",
    "import os\n",
    "from IPython.display import display, Markdown\n",
    "\n",
    "# Load the CSV file\n",
    "cwd = os.getcwd()\n",
    "cwd = os.path.dirname(cwd)\n",
    "gene_names = []\n",
    "\n",
    "\n",
    "def scan_gene_names(path):\n",
    "    dirs = [d for d in os.listdir(path) if os.path.isdir(os.path.join(path, d)) and d != \"filtered\"]\n",
    "    dirs.sort()\n",
    "    for dir in dirs:\n",
    "        amr_file = os.path.join(path, dir, dir + \"_assembly\", dir + \"_amrfinderplus_report.tsv\")\n",
    "        data = pd.read_csv(amr_file, sep=\"\\t\", dtype=str)\n",
    "        # get the values of the column \"Gene symbol\" in data, which is a dataframe\n",
    "        genes = data[\"Gene symbol\"].tolist()\n",
    "        # for each item in genes\n",
    "        for gene in genes:\n",
    "            if gene.lower() not in gene_names:\n",
    "                gene_names.append(gene.lower())\n",
    "\n",
    "\n",
    "def italic_gene_names_column(value: pd.Series) -> pd.Series:\n",
    "    # in each value of the Series, use italic_gene_names() to update it\n",
    "    return value.apply(italic_gene_names)\n",
    "\n",
    "\n",
    "def italic_gene_names(value: str) -> str:\n",
    "    # check every words in value, if it is in gene_names, make it italic\n",
    "    words = str(value).split()\n",
    "    for i, word in enumerate(words):\n",
    "        # if word in gene_names, ignore case\n",
    "        if word.lower() in gene_names:\n",
    "            words[i] = \"<i>\" + word + \"</i>\"\n",
    "    return \" \".join(words)\n",
    "\n",
    "\n",
    "def plot_table(data: pd.DataFrame, title: str):\n",
    "    fig = go.Figure(\n",
    "        data=[\n",
    "            go.Table(\n",
    "                header=dict(values=list(data.keys()), align=\"center\", font=dict(color=\"black\", size=14)),\n",
    "                cells=dict(\n",
    "                    values=[italic_gene_names_column(data[col]) for col in data],\n",
    "                    fill_color=\"lavender\",\n",
    "                    align=\"center\",\n",
    "                    font=dict(color=\"black\", size=14),\n",
    "                ),\n",
    "                columnwidth=[70, 70, 290, 70, 70, 70, 70],\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    title_text = f\"<div align='left' style='font-size:24px;'>{title}</div>\"\n",
    "    display(Markdown(title_text))\n",
    "\n",
    "    fig.update_layout(margin=dict(l=30, r=30, t=30, b=30))\n",
    "    fig.update_layout(width=1500, height=100 + 40 * len(data))\n",
    "    fig.show()\n",
    "\n",
    "\n",
    "path = \"Sanibel/output-date\"  # example input data path; you may change the path here\n",
    "scan_gene_names(path)\n",
    "\n",
    "# get all directories under path, and the directory name cannot be \"filtered\"\n",
    "dirs = [d for d in os.listdir(path) if os.path.isdir(os.path.join(path, d)) and d != \"filtered\"]\n",
    "dirs.sort()\n",
    "all_data = pd.DataFrame()\n",
    "for dir in dirs:\n",
    "    amr_file = os.path.join(path, dir, dir + \"_assembly\", dir + \"_amrfinderplus_report.tsv\")\n",
    "    data = pd.read_csv(amr_file, sep=\"\\t\", dtype=str)\n",
    "    data = data[\n",
    "        [\n",
    "            \"Gene symbol\",\n",
    "            \"Sequence name\",\n",
    "            \"Element type\",\n",
    "            \"Class\",\n",
    "            \"% Coverage of reference sequence\",\n",
    "            \"% Identity to reference sequence\",\n",
    "        ]\n",
    "    ]\n",
    "    # add a column \"Sample ID\" be the first column\n",
    "    data.insert(0, \"Sample ID\", dir)\n",
    "\n",
    "    # change column \"Gene symbol\" to \"Gene\"\n",
    "    data = data.rename(columns={\"Gene symbol\": \"Gene\"})\n",
    "    all_data = pd.concat([all_data, data], ignore_index=True)\n",
    "\n",
    "plot_table(all_data, \"Antimicrobial Resistance Genes\") # option 1: show the table in the notebook\n",
    "\n",
    "all_data.to_excel(\"AMR_genes.xlsx\", index=False) # option 2: save the table to an excel file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6. Geo map by county"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import os\n",
    "import json\n",
    "\n",
    "# Load the geojson file\n",
    "cwd = os.getcwd()\n",
    "with open(\"../Data/meta/fl-counties-fips.json\", \"r\") as f:  # update the path to the geojson file\n",
    "    geojson = json.load(f)\n",
    "\n",
    "# Load the CSV file\n",
    "meta = pd.read_csv(\"../Data/meta/meta_data.csv\") # example meta data file; you may change the meta data file here\n",
    "renamed_columns = [\"Sample ID\", \"Collection Date\", \"County\"]  # only show data for these columns; you may change the column names here\n",
    "meta.columns = renamed_columns\n",
    "\n",
    "\n",
    "def get_county_data(meta, geojson):\n",
    "    county_names = [\n",
    "        entry[\"properties\"][\"NAME\"]\n",
    "        for entry in geojson[\"features\"]\n",
    "        if entry[\"properties\"][\"STATE\"] == \"12\" # florida fips is 12\n",
    "    ]\n",
    "    county_counts = pd.DataFrame(county_names, columns=[\"County\"])\n",
    "    county_counts[\"FIPS\"] = [\n",
    "        entry[\"properties\"][\"GEO_ID\"][-5:] \n",
    "        for entry in geojson[\"features\"]\n",
    "        if entry[\"properties\"][\"STATE\"] == \"12\" # florida fips is 12\n",
    "    ]\n",
    "\n",
    "    counts = meta[\"County\"].value_counts().reset_index()\n",
    "\n",
    "    county_counts = county_counts.merge(counts, on=\"County\", how=\"left\")\n",
    "    county_counts[\"count\"].fillna(0, inplace=True)\n",
    "\n",
    "    return county_counts\n",
    "\n",
    "\n",
    "county_counts = get_county_data(meta, geojson)\n",
    "\n",
    "fig = px.choropleth(\n",
    "    county_counts,\n",
    "    geojson=geojson,\n",
    "    locations=\"FIPS\",\n",
    "    color=\"count\",\n",
    "    hover_name=\"County\",\n",
    "    scope=\"usa\",\n",
    "    title=\"Florida by County\",\n",
    "    color_continuous_scale=\"blues\",\n",
    ")\n",
    "\n",
    "fig.update_traces(marker_line=dict(color=\"rgba(0,0,0,0.3)\", width=0.2))\n",
    "fig.update_layout(autosize=True, margin=dict(t=0, b=0, l=0, r=0))\n",
    "fig.update_geos(fitbounds=\"locations\", visible=False)\n",
    "\n",
    "fig.update_layout(width=1500, height=1000)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7. Summary Varpipe-WGS results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.graph_objects as go\n",
    "import os\n",
    "from IPython.display import display, Markdown\n",
    "\n",
    "# Load the CSV file\n",
    "cwd = os.getcwd()\n",
    "cwd = os.path.dirname(cwd)\n",
    "gene_names = [\n",
    "    \"rpob\",\n",
    "    \"katg\",\n",
    "    \"inha\",\n",
    "    \"fabg1\",\n",
    "    \"embb\",\n",
    "    \"pnca\",\n",
    "    \"gyra\",\n",
    "    \"gyrb\",\n",
    "    \"mmpr\",  # \"rv0678 (mmpr)\",\n",
    "    \"atpe\",\n",
    "    \"rplc\",\n",
    "    \"rrl\",\n",
    "]\n",
    "\n",
    "\n",
    "def extract_interpretations(file_path):\n",
    "    with open(file_path, \"r\") as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    start_line = None\n",
    "    for i, line in enumerate(lines):\n",
    "        if line.strip() == \"Interpretations Summary:\":\n",
    "            start_line = i\n",
    "            break\n",
    "\n",
    "    if start_line is None:\n",
    "        raise ValueError(\"Interpretations Summary not found in the file\")\n",
    "\n",
    "    interpretations_data = []\n",
    "    headers = [\"Drug\", \"Variant\", \"Interpretation\"]\n",
    "\n",
    "    for line in lines[start_line + 2 :]:\n",
    "        if line.strip() == \"\":\n",
    "            break\n",
    "        interpretations_data.append(line.strip().split(\"\\t\"))\n",
    "\n",
    "    df = pd.DataFrame(interpretations_data, columns=headers)\n",
    "    return df\n",
    "\n",
    "\n",
    "def scan_target_files(path):\n",
    "    # scan all files under the path and the subdirectories to look for all files ended with \"_target_region_coverage.txt\"\n",
    "    target_files = {}\n",
    "    summary_files = {}\n",
    "    lineage_files = {}\n",
    "    for dirpath, dirnames, filenames in os.walk(path):\n",
    "        dirnames[:] = [d for d in dirnames if d != \"QC\"]\n",
    "        for filename in filenames:\n",
    "            sample_id = filename[:11]\n",
    "            if filename.endswith(\"_DR_loci_Final_annotation.txt\"):\n",
    "                target_files[sample_id] = os.path.join(dirpath, filename)\n",
    "            elif filename.endswith(\"_summary.txt\"):\n",
    "                summary_files[sample_id] = os.path.join(dirpath, filename)\n",
    "            elif filename.endswith(\"_Lineage.txt\"):\n",
    "                lineage_files[sample_id] = os.path.join(dirpath, filename)\n",
    "    return target_files, summary_files, lineage_files\n",
    "\n",
    "\n",
    "def italic_gene_names_column(value: pd.Series) -> pd.Series:\n",
    "    # in each value of the Series, use italic_gene_names() to update it\n",
    "    return value.apply(italic_gene_names)\n",
    "\n",
    "\n",
    "def italic_gene_names(value: str) -> str:\n",
    "    # check every words in value, if it is in gene_names, make it italic\n",
    "    words = str(value).split()\n",
    "    for i, word in enumerate(words):\n",
    "        # if word in gene_names, ignore case\n",
    "        if word.lower() in gene_names:\n",
    "            words[i] = \"<i>\" + word + \"</i>\"\n",
    "    return \" \".join(words)\n",
    "\n",
    "\n",
    "def plot_table_varpipe_target(target_file: str, summary_file, sample_id: str, lineage_file: str):\n",
    "    data = pd.read_csv(target_file, sep=\"\\t\", dtype=str)\n",
    "    summary_data = extract_interpretations(summary_file)\n",
    "    # for the column \"SAMPLE_ID\", only keep the first 11 characters\n",
    "    data[\"Sample ID\"] = data[\"Sample ID\"].str[:11]\n",
    "\n",
    "    # remove columns: CHROM, REF, ALT, TYPE, Position within CDS, REF Amino acid, ALT Amino acid, Codon Position, Gene ID\n",
    "    data = data.drop(\n",
    "        [\n",
    "            \"CHROM\",\n",
    "            \"REF\",\n",
    "            \"ALT\",\n",
    "            \"Variant Type\",\n",
    "            \"Position within CDS \",\n",
    "            \"REF Amino acid\",\n",
    "            \"ALT Amino acid\",\n",
    "            \"Codon Position\",\n",
    "            \"Gene ID\",\n",
    "        ],\n",
    "        axis=1,\n",
    "    )\n",
    "    # remove the rows that gene name is not in gene_names, ignore case\n",
    "    data = data[data[\"Gene Name\"].str.lower().isin(gene_names)]\n",
    "    # only keep the rows: 1. if 'Gene name' is gyrA or gyrB, 'Percent Alt Allele' should be at least 5, 2. for all other gene names, 'Percent Alt Allele' should be at least 10\n",
    "    data[\"Percent Alt Allele\"] = pd.to_numeric(data[\"Percent Alt Allele\"], errors=\"coerce\")\n",
    "    data = data[\n",
    "        (data[\"Gene Name\"].isin([\"gyrA\", \"gyrB\"]) & (data[\"Percent Alt Allele\"] >= 5))\n",
    "        | (~data[\"Gene Name\"].isin([\"gyrA\", \"gyrB\"]) & (data[\"Percent Alt Allele\"] >= 10))\n",
    "    ]\n",
    "\n",
    "    # remove the rows from data, if \"Amino Acid Change\" is \"p.Leu449Gln\" and \"Gene Name\" is \"rpoB\"\n",
    "    data = data[~((data[\"Amino acid Change\"] == \"p.Leu449Gln\") & (data[\"Gene Name\"] == \"rpoB\"))]\n",
    "\n",
    "    # add \"Interpretation\" from summary_data to data, only if the first few characters in \"Variant\" before \"_\" in summary_data is in the \"Gene Name\" in data\n",
    "    summary_data[\"Gene Name\"] = summary_data[\"Variant\"].str.split(\"_\").str[0]\n",
    "    data = data.merge(summary_data, on=\"Gene Name\", how=\"left\")\n",
    "    # remove the column \"Drug\" and \"Variant\" from data\n",
    "    data.drop([\"Drug\", \"Variant\"], axis=1, inplace=True)\n",
    "    # replace NaN with empty string in \"Interpretation\"\n",
    "    data[\"Interpretation\"] = data[\"Interpretation\"].fillna(\"\")\n",
    "\n",
    "    # we do not need interpretation any more requested by Calin\n",
    "    data = data.drop([\"Interpretation\"], axis=1)\n",
    "\n",
    "    # merge linage table to data\n",
    "    lineage = \"\"\n",
    "    long_sublineage = \"\"\n",
    "    with open(lineage_file, \"r\") as file:\n",
    "        sublineages = []\n",
    "        for line in file:\n",
    "            line = line.rstrip()\n",
    "            # if line starts with \"Lineage\"\n",
    "            if line.startswith(\"Lineage\"):\n",
    "                lineage = line[len(\"Lineage: \") : -1]\n",
    "            else:\n",
    "                if \"lineage\" in line:\n",
    "                    index = line.index(\"lineage\") + len(\"lineage\") + 1\n",
    "                    sublineages.append(line[index : len(line)].strip())\n",
    "                elif \"suggests\" in line:\n",
    "                    index = line.index(\"suggests\") + len(\"suggests\") + 1\n",
    "                    sublineages.append(line[index : len(line)].strip())\n",
    "        long_sublineage = max(sublineages, key=len)\n",
    "\n",
    "    data[\"Lineage\"] = lineage\n",
    "    data[\"Sub-lineage\"] = long_sublineage\n",
    "\n",
    "    return data\n",
    "\n",
    "\n",
    "def plot_table_lineage(lineage_file, sample_id):\n",
    "    with open(lineage_file, \"r\") as file:\n",
    "        sublineages = []\n",
    "        for line in file:\n",
    "            line = line.rstrip()\n",
    "            # if line starts with \"Lineage\"\n",
    "            if line.startswith(\"Lineage\"):\n",
    "                lineage = line[len(\"Lineage: \") : -1]\n",
    "            else:\n",
    "                if \"lineage\" in line:\n",
    "                    index = line.index(\"lineage\") + len(\"lineage\") + 1\n",
    "                    sublineages.append(line[index : len(line)].strip())\n",
    "                elif \"suggests\" in line:\n",
    "                    index = line.index(\"suggests\") + len(\"suggests\") + 1\n",
    "                    sublineages.append(line[index : len(line)].strip())\n",
    "        long_sublineage = max(sublineages, key=len)\n",
    "\n",
    "    # new a dataframe with three columns: \"Sample ID\", \"Lineage\", \"Sub-lineage\"\n",
    "    data = pd.DataFrame(columns=[\"Sample ID\", \"Lineage\", \"Sub-lineage\"])\n",
    "    # add a new row with value of sample_id, lineage, sublineage, respectivaly\n",
    "    data.loc[0] = [sample_id, lineage, long_sublineage]\n",
    "\n",
    "    fig = go.Figure(\n",
    "        data=[\n",
    "            go.Table(\n",
    "                header=dict(\n",
    "                    values=list(data.keys()),\n",
    "                    align=\"center\",\n",
    "                    font=dict(color=\"black\", size=14),\n",
    "                ),\n",
    "                cells=dict(\n",
    "                    values=[italic_gene_names_column(data[col]) for col in data],\n",
    "                    fill_color=\"lavender\",\n",
    "                    align=\"center\",\n",
    "                    font=dict(color=\"black\", size=14),\n",
    "                    height=25,\n",
    "                ),\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    fig.update_layout(margin=dict(l=30, r=30, t=30, b=30))\n",
    "    fig.update_layout(width=600, height=100 + len(data) * 32)\n",
    "    fig.show()\n",
    "\n",
    "\n",
    "\n",
    "meta_data = pd.read_csv(\"meta.csv\")  # add the path here\n",
    "# meta_data.drop(columns=[\"SOURCE\", \"Date of receipt\", \"Submitter\"], inplace=True)\n",
    "meta_data.columns = [\"Sample ID\", \"Collection Date\", \"County\"]\n",
    "\n",
    "\n",
    "path = \"path/to/fastqs_varpipe\"  # add the path here\n",
    "target_files, summary_files, lineage_files = scan_target_files(path)\n",
    "\n",
    "gene_names = {item.lower() for item in gene_names}\n",
    "\n",
    "total_data = pd.DataFrame()\n",
    "for sample_id in meta_data[\"Sample ID\"].tolist():\n",
    "    # if sample_id is a key in the target_files\n",
    "    if sample_id in target_files:\n",
    "        target_file = target_files[sample_id]\n",
    "        summary_file = summary_files[sample_id]\n",
    "        data = plot_table_varpipe_target(target_file, summary_file, sample_id, lineage_files[sample_id])\n",
    "        total_data = pd.concat([total_data, data])\n",
    "\n",
    "def plot_table_total(total_data):\n",
    "    fig = go.Figure(\n",
    "        data=[\n",
    "            go.Table(\n",
    "                header=dict(\n",
    "                    values=list(total_data.keys()),\n",
    "                    align=\"center\",\n",
    "                    font=dict(color=\"black\", size=14),\n",
    "                ),\n",
    "                cells=dict(\n",
    "                    values=[italic_gene_names_column(total_data[col]) for col in total_data],\n",
    "                    fill_color=\"lavender\",\n",
    "                    align=\"center\",\n",
    "                    font=dict(color=\"black\", size=14),\n",
    "                    height=25,\n",
    "                ),\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    fig.update_layout(margin=dict(l=30, r=30, t=30, b=30))\n",
    "    fig.update_layout(width=1500, height=100 + len(total_data) * 32)\n",
    "    fig.show()\n",
    "\n",
    "if len(total_data) < 100: # if the number of rows in total_data is less than 100, plot the table\n",
    "    plot_table_total(total_data)\n",
    "else: # if the number of rows in total_data is greater than 100, save the table to a csv file\n",
    "    total_data.to_csv(\"Table - Genetic modifications.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [
     "exclude"
    ]
   },
   "source": [
    "### 8. To generate a report in html format, run the following command. \n",
    "#### - Make sure you change the .ipynb file name to the one you are preparing.\n",
    "#### - Make sure you have nbconvert installed in your conda environment\n",
    "#### - Add the tag \"exclude\" to the cells which you do not want to show in the generated html file\n",
    "\n",
    "\n",
    "       `jupyter nbconvert --execute summary.ipynb --to html --no-prompt --TagRemovePreprocessor.remove_cell_tags \"exclude\" --TemplateExporter.exclude_input=True`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
